x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/latest/LatestDoc.java

```diff
@@ -0,0 +1,280 @@
+
+import static java.util.function.Predicate.not;
+import static java.util.stream.Collectors.toMap;
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+
+public class LatestDoc implements Function {
+
+    public static final int DEFAULT_INITIAL_MAX_PAGE_SEARCH_SIZE = 5000;
+    public static final int TEST_QUERY_PAGE_SIZE = 50;
+
+    private static final String COMPOSITE_AGGREGATION_NAME = "_transform";
+    private static final String TOP_HITS_AGGREGATION_NAME = "_top_hits";
+    private static final Logger logger = LogManager.getLogger(LatestDoc.class);
+
+    private final LatestDocConfig config;
+
+    // objects for re-using
+    private final CompositeAggregationBuilder cachedCompositeAggregation;
+
+    public LatestDoc(LatestDocConfig config) {
+        this.config = config;
+        this.cachedCompositeAggregation = createCompositeAggregation(config);
+    }
+
+    private static CompositeAggregationBuilder createCompositeAggregation(LatestDocConfig config) {
+        assert config.getSort().size() == 1 && config.getSort().get(0) instanceof FieldSortBuilder;
+
+        CompositeAggregationBuilder compositeAggregation;
+        try (XContentBuilder builder = jsonBuilder()) {
+            config.toCompositeAggXContent(builder);
+            XContentParser parser = builder.generator()
+                .contentType()
+                .xContent()
+                .createParser(NamedXContentRegistry.EMPTY, LoggingDeprecationHandler.INSTANCE, BytesReference.bytes(builder).streamInput());
+            compositeAggregation = CompositeAggregationBuilder.PARSER.parse(parser, COMPOSITE_AGGREGATION_NAME);
+        } catch (IOException e) {
+            throw new RuntimeException(
+                TransformMessages.getMessage(TransformMessages.TRANSFORM_FAILED_TO_CREATE_COMPOSITE_AGGREGATION, "latest_doc"), e);
+        }
+        TopHitsAggregationBuilder topHitsAgg =
+            AggregationBuilders.topHits(TOP_HITS_AGGREGATION_NAME)
+                .size(1)  // we are only interested in the top-1
+                .sorts(config.getSort());  // we copy the sort config directly from the function config
+        compositeAggregation.subAggregation(topHitsAgg);
+        return compositeAggregation;
+    }
+
+    @Override
+    public int getInitialPageSize() {
+        return DEFAULT_INITIAL_MAX_PAGE_SEARCH_SIZE;
+    }
+
+    private SearchRequest buildSearchRequest(SourceConfig sourceConfig, Map<String, Object> position, int pageSize) {
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
+        buildSearchQuery(sourceBuilder, null, pageSize);
+        sourceBuilder.query(sourceConfig.getQueryConfig().getQuery());
+        SearchRequest searchRequest =
+            new SearchRequest(sourceConfig.getIndex())
+                .source(sourceBuilder)
+                .indicesOptions(IndicesOptions.LENIENT_EXPAND_OPEN);
+        logger.trace("Search request: {}", searchRequest);
```

Comment:
```suggestion
        logger.trace(() -> new ParameterizedMessage("Search request: {}", searchRequest));
```

Reply:
Done.

Reply:
as a result of my other comment, I wonder if this trace is useless. We have much better logging in the indexer and that's what people should use for debugging a transform problem. This is only called in validation and preview.